oauth/src/lib.rs

```diff
@@ -162,6 +213,208 @@ fn get_socket_address(redirect_uri: &str) -> Option<SocketAddr> {
     None
 }
 
+/// Struct that handle obtaining and refreshing access tokens.
+pub struct OAuthClient {
+    scopes: Vec<String>,
+    redirect_uri: String,
+    should_open_url: bool,
+    message: String,
+    client: BasicClient,
+}
+
+impl OAuthClient {
+    /// Generates and opens/shows the authorization URL to obtain an access token.
+    ///
+    /// Returns a verifier that must be included in the final request for validation.
+    fn set_auth_url(&self) -> PkceCodeVerifier {
+        let (pkce_challenge, pkce_verifier) = PkceCodeChallenge::new_random_sha256();
+        // Generate the full authorization URL.
+        // Some of these scopes are unavailable for custom client IDs. Which?
+        let request_scopes: Vec<oauth2::Scope> =
+            self.scopes.iter().map(|s| Scope::new(s.into())).collect();
+        let (auth_url, _) = self
+            .client
+            .authorize_url(CsrfToken::new_random)
+            .add_scopes(request_scopes)
+            .set_pkce_challenge(pkce_challenge)
+            .url();
+
+        if self.should_open_url {
+            open::that_in_background(auth_url.as_str());
+        } else {
+            println!("Browse to: {}", auth_url);
+        }
```

Comment:
Sorry, I missed in my earlier review that this is different logic to what was proposed in #1396. Here, you don't print `auth_url` in the case where opening was configured. But I think we should always print it out. It costs nothing and avoids the whole thing being unusable when run on a system without a browser. 

```suggestion
        if self.should_open_url {
            open::that_in_background(auth_url.as_str());
        }
        println!("Browse to: {}", auth_url);
```